ocaml/typing/typeopt.ml

```diff
@@ -181,29 +185,78 @@ let classify env loc ty sort : classification =
   | Base Word -> Unboxed_int Pnativeint
   | Base Void as c ->
     raise (Error (loc, Unsupported_sort c))
-  | Product c -> Product c
+  | Product c -> Product (classify_product ty c)
+
+let rec scannable_product_array_kind loc sorts =
+  List.map (sort_to_scannable_product_element_kind loc) sorts
+
+and sort_to_scannable_product_element_kind loc (s : Jkind.Sort.Const.t) =
+  (* Unfortunate: this never returns `Pint_scannable`.  Doing so would require
```

Comment:
> Indeed. This seems kinda bad.

I think there is an interesting discussion to be had here, but it's probably worth saying explicitly that that discussion shouldn't hold up this PR. The thing implemented here is fine for a first version.

> We could, of course, introduce product kinds. But I continue to think this is the wrong idea -- and it will only get worse, once kinds have `with` things in them.
> 
> What I want is a way to store both a sort and a gc-ignore-flag (gif) for each component of a product. But if we have a subsumption relationship between can-ignore and cannot-ignore, then we lose our nice property of having a discrete set of incomparable sorts. After thinking about this a little while, I'm stumped. Let's talk about this.

I remain confused why we wouldn't want product kinds. The more time we spend working around their absence, the more I think we should at least give them a try. It's definitely an interesting observation that we can work around not having them, but I don't think they complicate inference, and in fact I think they will just make the whole system simpler.

I'm certainly not proposing to introduce them right now, though - we should get this and a first version of unboxed records and interior pointers out with the system we have.

(Speaking of making the whole system simpler, I think can even imagine a world where product kinds allow us to eliminate product both product layouts and product sorts. Though I think they'd simplify the system even if not.)

> (To be clear, even if we had product kinds, I think the inference problem lurks. So it's just not as simple as "add product kinds".)

This is not clear to me. I think product kinds would lead to a straightforward implementation of the product cases of `array_kind_of_elt`, working only on the kind.  We would (1) classify each element of the product by its layout and ignorability, (2) check we don't have a mixed block, and (3) return the obvious array kind.

> Actually, I've thought about this more (longer than I should admit, but in the evening at least).
> 
> I think trying to infer stuff in this space is the wrong thing to be trying for. We don't, for example, try to eagerly figure out where we can specialize types to `int` so that typeopt can do its thing. Instead, here in typeopt, we look deeply through types to figure out all we can.

I agree that typeopt looks deeply through types.

I am not convinced this is something to emulate, though. `value_kind` is a significant source of both complexity and performance problems, exactly because of this. I continue to think there should be one function that answers the question "what is the kind of this type", and that function is `estimate_type_jkind`.

I can actually imagine wanting to go further the other way - fold more of what `value_kind` is looking for into the kind system, until its implementation too can just look at the kind.

> (EDIT: The next bit of text probably is more about @rtjoa's unboxed records than this current patch, but it all comes back together in the end, and I know you @ccasin care deeply about both patches.)
> 
> And, thinking about "look deeply through types" made me realize what the specification of `scrape_ty` is: `scrape_ty env ty` returns a `ty'` whose runtime representation is the same as `ty` (but is somehow more primitive / easier to understand).

I agree that's a reasonable specification of `scrape_ty`. `scrape_ty` is just calls to `expand_head_opt` and `get_unboxed_type_approximation`, and this is the specification of those functions (or, at least, that has been my interpretation of the somewhat cryptic comment about "type-based optimizations" on `expand_head_opt`).

> This suggests that we should add a new ability to `scrape_ty`: "look through" an unboxed record to get an unboxed tuple. This meets the specification of `scrape_ty` beautifully: an unboxed tuple shares a runtime representation with an unboxed record, and the tuple is easier to process. (Yes, I know I just introduced that specification for `scrape_ty` and then used it -- but I really do think this is the useful specification of that function.)

You may recall from an earlier conversation that @rtjoa's first prototype for unboxed records took this approach (having `get_unboxed_type_representation` return an unboxed tuple when it sees an unboxed record).  I think this is a fun idea, but doesn't work here because of mutability. Note that `type_opt` cares about mutability of record fields in `value_kind`, and I expect this approach would make it hard to implement that logic when we allow for nested unboxed records in other structures.

(One could imagine enhancing tuple types to add a notion of mutable field just for this purpose, but for the reasons above I'm not particularly inclined to go this route.)

> We can go further. I conjecture that, today, `scrape_ty` and `get_unboxed_type_approximation` always do the same thing (though the latter does it with a little less work). (Modulo correcting levels, of course. That step, happily, does not need to be done once we merge with 5.3.)

It's a nice observation that `scrape_ty` and `get_unboxed_type_approximation` have pretty similar implementations and could potentially be just one function.  I think that might be right.

Reply:
> I think there is an interesting discussion to be had here, but it's probably worth saying explicitly that that discussion shouldn't hold up this PR. The thing implemented here is fine for a first version.

Strong agree.

> You may recall from an earlier conversation that @rtjoa's first prototype for unboxed records took this approach (having get_unboxed_type_representation return an unboxed tuple when it sees an unboxed record).

Yes, I do! And I should have noted this in my post; sorry for the oversight.

As for the rest: this will be much better discussed live than on GitHub. It would be reasonable to delay that discussion until after this all lands, but I think it's all fresh now and worth at least agreeing on next steps here. I'll suggest a time, but I also understand if we want to delay.